File: src/ai_team/ui/web/frontend/src/utils/compareSummary.ts
import type { MonitorState } from "../types";

type MetricExtractor = (m: MonitorState) => number | string;

/** Lower is better for elapsed; higher is better for counts. */
export function bestColumnKey(
  rows: { key: string; m: MonitorState }[],
  extract: MetricExtractor,
  prefer: "min" | "max",
): string | null {
  if (rows.length === 0) return null;
  let best = rows[0];
  let bestVal = extract(best.m);
  for (const r of rows.slice(1)) {
    const v = extract(r.m);
    const vn = Number(v);
    const bn = Number(bestVal);
    let better: boolean;
    if (!Number.isNaN(vn) && !Number.isNaN(bn)) {
      better = prefer === "min" ? vn < bn : vn > bn;
    } else {
      const cmp = String(v).localeCompare(String(bestVal), undefined, { numeric: true });
      better = prefer === "min" ? cmp < 0 : cmp > 0;
    }
    if (better) {
      best = r;
      bestVal = v;
    }
  }
  return best.key;
}

export function parseElapsedSeconds(elapsed: string): number {
  const m = elapsed.match(/(\d+)m/);
  const s = elapsed.match(/(\d+)s/);
  return (m ? parseInt(m[1], 10) * 60 : 0) + (s ? parseInt(s[1], 10) : 0);
}

export interface VerdictMetric {
  label: string;
  prefer: "min" | "max";
  numeric: (m: MonitorState) => number;
}

/** Build a one-line auto verdict from summary rows and metrics. */
export function buildCompareVerdict(
  rows: { key: string; label: string; m: MonitorState; failed?: boolean }[],
  metrics: VerdictMetric[],
): string {
  const active = rows.filter((r) => !r.failed);
  if (active.length === 0) return "";
  const parts: string[] = [];
  for (const metric of metrics) {
    const bestKey = bestColumnKey(active, metric.numeric, metric.prefer);
    if (!bestKey) continue;
    const winner = active.find((r) => r.key === bestKey);
    if (winner) {
      const direction = metric.prefer === "min" ? "lowest" : "highest";
      parts.push(`${winner.label}: ${direction} ${metric.label.toLowerCase()}`);
    }
  }
  return parts.join("; ");
}

export function directionHint(prefer: "min" | "max" | null): string {
  if (prefer === "min") return "▼ lower is better";
  if (prefer === "max") return "▲ higher is better";
  return "";
}

File: src/ai_team/ui/compare_summary.py
"""Compare summary helpers shared by web-facing Python clients (TUI)."""

from __future__ import annotations

import re
from typing import Any, Callable, Literal


def parse_elapsed_seconds(elapsed: str) -> int:
    """Parse monitor elapsed strings like ``2m 30s`` into seconds."""
    minutes = re.search(r"(\d+)m", elapsed or "")
    seconds = re.search(r"(\d+)s", elapsed or "")
    return (int(minutes.group(1)) * 60 if minutes else 0) + (
        int(seconds.group(1)) if seconds else 0
    )


def best_column_key(
    rows: list[dict[str, Any]],
    extract: Callable[[dict[str, Any]], int | float],
    prefer: Literal["min", "max"],
) -> str | None:
    """Return the row key with the best numeric metric."""
    if not rows:
        return None
    best = rows[0]
    best_val = extract(best)
    for row in rows[1:]:
        val = extract(row)
        if prefer == "min" and val < best_val:
            best, best_val = row, val
        elif prefer == "max" and val > best_val:
            best, best_val = row, val
    return str(best.get("key"))


def build_compare_verdict(
    rows: list[dict[str, Any]],
    metrics: list[tuple[str, Literal["min", "max"], Callable[[dict[str, Any]], float]]],
) -> str:
    """Build a one-line verdict string from summary rows."""
    active = [r for r in rows if not r.get("failed")]
    parts: list[str] = []
    for label, prefer, extract in metrics:
        best_key = best_column_key(active, extract, prefer)
        if not best_key:
            continue
        winner = next((r for r in active if r.get("key") == best_key), None)
        if winner:
            direction = "lowest" if prefer == "min" else "highest"
            parts.append(f"{winner.get('label', best_key)}: {direction} {label}")
    return "; ".join(parts)


def monitor_metric_rows(monitor: dict[str, Any]) -> dict[str, Any]:
    """Extract comparable metrics from a serialized monitor snapshot."""
    metrics = monitor.get("metrics") or {}
    return {
        "elapsed": monitor.get("elapsed", "—"),
        "elapsed_sec": parse_elapsed_seconds(str(monitor.get("elapsed", ""))),
        "phase": monitor.get("phase", "—"),
        "cost_usd": monitor.get("cost_usd"),
        "tokens": monitor.get("token_estimate") or 0,
        "tasks_completed": metrics.get("tasks_completed", 0),
        "tasks_failed": metrics.get("tasks_failed", 0),
        "files": metrics.get("files_generated", 0),
        "guardrails_passed": metrics.get("guardrails_passed", 0),
        "guardrails_failed": metrics.get("guardrails_failed", 0),
        "tests_passed": metrics.get("tests_passed", 0),
        "tests_failed": metrics.get("tests_failed", 0),
        "retries": metrics.get("retries", 0),
    }
